rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -52,10 +56,100 @@ NodeParameters::NodeParameters(
     use_intra_process,
     allocator);
 
+  // Get the node options
+  const rcl_node_t * node = node_base->get_rcl_node_handle();
+  if (nullptr == node) {
+    throw std::runtime_error("Need valid node handle in NodeParameters");
+  }
+  const rcl_node_options_t * options = rcl_node_get_options(node);
+  if (nullptr == options) {
+    throw std::runtime_error("Need valid node options NodeParameters");
+  }
+
+  // Get paths to yaml files containing initial parameter values
+  std::vector<std::string> yaml_paths;
+
+  auto get_yaml_paths = [&yaml_paths, &options](const rcl_arguments_t * args) {
+      int num_yaml_files = rcl_arguments_get_param_files_count(args);
+      if (num_yaml_files > 0) {
+        char ** param_files;
+        rcl_ret_t ret = rcl_arguments_get_param_files(args, options->allocator, &param_files);
+        if (RCL_RET_OK != ret) {
+          rclcpp::exceptions::throw_from_rcl_error(ret);
+        }
+        auto cleanup_param_files = make_scope_exit(
+          [&param_files, &num_yaml_files, &options]() {
+            for (int i = 0; i < num_yaml_files; ++i) {
+              options->allocator.deallocate(param_files[i], options->allocator.state);
+            }
+            options->allocator.deallocate(param_files, options->allocator.state);
+          });
+        for (int i = 0; i < num_yaml_files; ++i) {
+          yaml_paths.emplace_back(param_files[i]);
+        }
+      }
+    };
+
+  // global before local so that local overwrites global
+  if (options->use_global_arguments) {
+    get_yaml_paths(rcl_get_global_arguments());
+  }
+  get_yaml_paths(&(options->arguments));
+
+  // Get fully qualified node name post-remapping to use to find node's params in yaml files
+  const std::string node_name = node_base->get_name();
+  const std::string node_namespace = node_base->get_namespace();
+  if (0u == node_namespace.size() || 0u == node_name.size()) {
+    // Should never happen
+    throw std::runtime_error("Node name and namespace were not set");
+  }
+  std::string combined_name;
+  if ('/' == node_namespace.at(node_namespace.size() - 1)) {
+    combined_name = node_namespace + node_name;
+  } else {
+    combined_name = node_namespace + '/' + node_name;
+  }
+
+  std::map<std::string, rclcpp::Parameter> parameters;
+
+  // TODO(sloretz) use rcl to parse yaml when circular dependency is solved
```

Comment:
@sloretz do you happen to remember the context around this todo? I came across it today and was confused since it seems like yaml parsing is done in rcl already, so it must be referring to something else.

Reply:
> since it seems like yaml parsing is done in rcl already

This is actually done in rcl_yaml_param_parser and not rcl.
That package is currently depending on rcl to be able to use rcl types, so rcl cannot depenc on rcl_yaml_param_parser to avoid circular dependency.
https://github.com/ros2/rcl/issues/252

Reply:
oh I see! Thanks, I thought from the `rcl_` prefix that they were in rcl, but of course that's not the case :smile: I'll reference that ticket in the todo.